### PR TITLE
Fix #26: Fetch expected value from server prior to fetching records

### LINKED ISSUE #26: Get _expected value from server
Instead of requiring the `_expected` parameter on `.get()`, we could obtain it from the server (more convenient for the user).

* Fetch value of `last_modified` from `${server}/buckets/monitor/collections/changes/records?collection=${collection}&bucket=${bucket}`
* Use this value as the `_expected` parameter when fetching from the `.../changeset` endpoint
* Remove the `expected` parameter from `.get()`

### PR DESCRIPTION
**Pull request by Vishwa-Mozilla**

### CODE REVIEW

File: src/client/kinto_http.rs [hunk 1]
```diff
@@ -48,18 +58,19 @@ impl From<std::num::ParseIntError> for KintoError {
     }
 }
 
-pub async fn get_records(
+pub async fn get_last_modified_change(
```
Comment (leplatrem) [inline src/client/kinto_http.rs:61]: I liked your naming for `LatestChangeResponse`. Maybe this could be `get_latest_change_timestamp` ?

File: src/client/kinto_http.rs [hunk 2]
```diff
@@ -80,6 +100,9 @@ pub async fn get_changeset(
     cid: &str,
     expected: u64,
 ) -> Result<ChangesetResponse, KintoError> {
+
+    debug!("The expected value for bucket={}, collection={} is {}", bid, cid, expected);
```
Comment (leplatrem) [inline src/client/kinto_http.rs:104]: *The expected timestamp for...*

File: src/client/kinto_http.rs [hunk 3]
```diff
@@ -68,10 +79,19 @@ pub async fn get_records(
...
+    
     let body = resp.text().await?;
-    let result: ChangesetResponse = serde_json::from_str(&body)?;
 
-    Ok(result)
+    let latest_change: KintoPluralResponse<LatestChangeResponse> = serde_json::from_str(&body)?;
+    
+    let last_modified = match latest_change.data.get(0) {
+        Some(change) => change.last_modified,
+        _ => 0 // 0 by default
```
Comment (leplatrem) [inline src/client/kinto_http.rs:89]: We should raise an error here. If there is no entry in this endpoint, it means the bucket/collection is unknown. The call to changeset will raise a 404, so we better fail early

File: src/client.rs [hunk 4]
```diff
@@ -223,85 +223,109 @@ mod tests {
...
+            debug!("test name: {}", test.test_description);
+
+            let mut get_latest_change_mock = Mock::new()
+            .expect_method(GET)
+            .expect_path("/buckets/monitor/collections/changes/records")
+            .expect_query_param("bucket", "main")
+            .expect_query_param("collection", "url-classifier-skip-urls")
+            .return_status(200)
+            .return_header("Content-Type", "application/json")
+            .return_header("etag", "12345")
```
Comment (leplatrem) [inline src/client.rs:298]: Is this used somewhere?

File: src/client.rs [hunk 5]
```diff
@@ -223,85 +223,109 @@ mod tests {
...
+            .return_body(
             r#"{
-                "metadata": {},
-                "changes": [],
-                "timestamp": 0
+                "data": [
+                    {
+                        "id": "123",
+                        "last_modified": 9173,
```
Comment (leplatrem) [inline src/client.rs:304]: why not use `EXPECTED_VERSION` here?

File: src/client.rs [hunk 6]
```diff
@@ -223,85 +223,109 @@ mod tests {
...
+            .return_body(test.changeset_response_body).create_on(&mock_server);
+
+            let actual_result = test.client.get().await;
+            assert_eq!(actual_result, test.expected_result);
+            assert_eq!(1, get_changeset_mock.times_called());
+            assert_eq!(1, get_latest_change_mock.times_called());
+
+            get_changeset_mock.delete();
+            get_latest_change_mock.delete();
+        }
```
Comment (leplatrem) [inline src/client.rs:328]: You should add a test here to assert the behaviour when bucket/collection are unknown on the server

File: src/client/kinto_http.rs (hunk 1 above)
Comment (Vishwa-Mozilla) [reply src/client/kinto_http.rs:61]: done

File: src/client/kinto_http.rs (hunk 2 above)
Comment (Vishwa-Mozilla) [reply src/client/kinto_http.rs:104]: done

File: src/client.rs (hunk 4 above)
Comment (Vishwa-Mozilla) [reply src/client.rs:298]: to get the timestamp from the response

File: src/client/kinto_http.rs (hunk 3 above)
Comment (Vishwa-Mozilla) [reply src/client/kinto_http.rs:89]: done

File: src/client.rs (hunk 4 above)
Comment (leplatrem) [reply src/client.rs:298]: and what do we do with it? I don't think it's necessary

Comment (Vishwa-Mozilla) [reply src/client.rs:298]: its used for debugging but nothing other than that - I can remove it

File: src/client.rs (hunk 5 above)
Comment (Vishwa-Mozilla) [reply src/client.rs:304]: yes

File: src/client.rs (hunk 6 above)
Comment (Vishwa-Mozilla) [reply src/client.rs]: done

File: rs-client-demo/src/main.rs [hunk 7]
```diff
@@ -29,133 +29,131 @@ fn main() {
...
-    //        Ok(records) => println!("Successful fetch"),
-    //        Err(error) => {
-    //            println!("Could not fetch records: {:?}", error);
-    //            failed_fetch.push(collection);
-    //        },
-    //    };
+       match client.get() {
+           Ok(records) => println!("Successful fetch"),
+           Err(error) => {
+               println!("Could not fetch records: {:?}", error);
```
Comment (leplatrem) [inline rs-client-demo/src/main.rs:140]: It's not only about fetching. Signature verification can fail, right?

File: src/client/kinto_http.rs [hunk 8]
```diff
@@ -79,9 +78,23 @@ pub fn get_records(
...
-    let result: ChangesetResponse = resp.json()?;
 
-    Ok(result)
+    let latest_change: KintoPluralResponse<LatestChangeResponse> = resp.json()?;
+    
+    let last_modified = match latest_change.data.get(0) {
+        Some(change) => change.last_modified,
+        None => {
+            // bucket/collection provided is unknown
+            return Err(KintoError::Error {name: format!("404 - unknown bucket/collection name provided: bucket_name={}, collection_name{}", bid, cid)});
```
Comment (leplatrem) [inline src/client/kinto_http.rs:88]: I don't think 404 makes sense here. Error can just be `format!("Unknown collection {}/{}", bid, cid)`

File: src/client/kinto_http.rs [hunk 9]
```diff
@@ -90,6 +103,11 @@ pub fn get_changeset(
     cid: &str,
     expected: u64,
 ) -> Result<ChangesetResponse, KintoError> {
+    debug!(
+        "The expected timestamp for bucket={}, collection={} is {}",
```
Comment (leplatrem) [inline src/client/kinto_http.rs:107]: in most places (and other clients), we refer as a collection like this : `{bid}/{cid}`

File: src/client/signatures/openssl_verifier.rs [hunk 10]
```diff
@@ -144,15 +144,11 @@ impl Verification for OpenSSLVerifier {
...
-    use super::{Collection, OpenSSLVerifier, SignatureError, Verification};
+    use super::{Collection, OpenSSLVerifier, Verification};
     use env_logger;
     use httpmock::Method::GET;
     use httpmock::{Mock, MockServer};
-    use log::debug;
     use serde_json::json;
-    use url::{ParseError, Url};
-    use viaduct::{set_backend, Error as ViaductError, Request};
-    use viaduct_reqwest::ReqwestBackend;
```
Comment (leplatrem) [inline src/client/signatures/openssl_verifier.rs:155]: This change seems unrelated

File: src/client.rs [hunk 11]
```diff
@@ -177,169 +174,219 @@ impl Client {
...
+        let mock_server_address = mock_server.url("");
+
+        let valid_latest_change_response = &format!(
+            "{}{}{}",
+            r#"{
+            "data": [
+                {
+                    "id": "123",
+                    "last_modified": "#,
+            EXPECTED_VERSION,
```
Comment (leplatrem) [inline src/client.rs:300]: since `EXPECTED_VERSION` is not passed as parameter in `test_get()` this is kind of useless (and a bit ugly TBH)

File: src/client.rs [hunk 12]
```diff
@@ -177,169 +174,219 @@ impl Client {
...
+            ]
+        }"#
+        );
+
+        let invalid_latest_change_response = &format!(
+            "{}",
+            r#"{
+            "data": []
+        }"#
+        );
```
Comment (leplatrem) [inline src/client.rs:315]: This is defined too far from its usage. A variable may not be necessary for that

File: src/client.rs [hunk 13]
```diff
@@ -177,169 +174,219 @@ impl Client {
...
+            "url-classifier-skip-urls",
+            Some(Box::new(VerifierWithNoError {})),
+        ), invalid_latest_change_response, r#"{
+            "metadata": {
+                "data": "test"
+            },
+            "changes": [{
+                "id": 1,
+                "last_modified": 100
+            }],
```
Comment (leplatrem) [inline src/client.rs:388]: We don't content in `changes` for this test case

File: src/client.rs [hunk 14]
```diff
@@ -177,169 +174,219 @@ impl Client {
...
+                Some(Box::new(VerifierWithVerificatonError {})),
+            ),
+            valid_latest_change_response,
+            r#"{
+            "metadata": {},
+            "changes": [],
+            "timestamp": 0
+        }"#,
+            Err(ClientError::VerificationError {
+                name: "verification error".to_owned(),
```
Comment (leplatrem) [inline src/client.rs:331]: `signature verification error` would be clearer IMO

File: src/client/kinto_http.rs (hunk 8 above)
Comment (Vishwa-Mozilla) [reply src/client/kinto_http.rs:88]: done

File: src/client.rs (hunk 14 above)
Comment (Vishwa-Mozilla) [reply src/client.rs:331]: done

File: rs-client-demo/Cargo.toml [hunk 15]
```diff
@@ -5,10 +5,6 @@ authors = ["Vishwa Patel <vpatel@mozilla.com>"]
 edition = "2018"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-[dev-dependencies]
-# cannot use tags since viaduct, viaduct-reqwest is not published yet  
-viaduct = { git = "https://github.com/mozilla/application-services", rev = "61dcc364ac0d6d0816ab88a494bbf20d824b009b"}
-viaduct-reqwest = { git = "https://github.com/mozilla/application-services", rev = "61dcc364ac0d6d0816ab88a494bbf20d824b009b"}
```
[the remark this answers is not in the document]
Comment (Vishwa-Mozilla) [reply rs-client-demo/Cargo.toml:11]: alright

File: src/client/kinto_http.rs (hunk 9 above)
Comment (Vishwa-Mozilla) [reply src/client/kinto_http.rs:107]: done

File: rs-client-demo/src/main.rs (hunk 7 above)
Comment (Vishwa-Mozilla) [reply rs-client-demo/src/main.rs:140]: yes

Comment (leplatrem) [reply rs-client-demo/src/main.rs:140]: Then we should adjust the message ;)

Comment (Vishwa-Mozilla) [reply rs-client-demo/src/main.rs:140]: agreed

File: src/client.rs (hunk 11 above)
Comment (Vishwa-Mozilla) [reply src/client.rs:300]: done

File: rs-client-demo/src/main.rs [hunk 16]
```diff
@@ -29,133 +29,131 @@ fn main() {
...
+            ("main", "sites-classification"),
+            ("main-preview", "sites-classification"),
+            ("main", "rocket-prefs"),
+            ("main-preview", "rocket-prefs"),
+            ("blocklists-preview", "gfx"),
+            ("blocklists", "qa"),
+            ("pinning-preview", "pins"),
+            ("pinning", "pins"),
+            ("blocklists", "gfx")
+        ];
```
[the remark this answers is not in the document]
Comment (Vishwa-Mozilla) [reply rs-client-demo/src/main.rs:125]: done

File: rs-client-demo/src/main.rs [hunk 17]
```diff
@@ -3,11 +3,26 @@
...
 use env_logger;
 use remote_settings_client::{Client, Collection, SignatureError, Verification};
-pub use viaduct::{set_backend};
+use viaduct::{set_backend, Request};
+pub use url::{ParseError, Url};
 pub use viaduct_reqwest::ReqwestBackend;
 use serde_json;
+use serde::Deserialize;
 
 struct CustomVerifier {}
```
[the remark this answers is not in the document]
Comment (Vishwa-Mozilla) [reply rs-client-demo/src/main.rs:12]: done
